Scrape mlab nodes.

pboothe opened the pull request:
Initial code to scrape mlab nodes.  Doesn't (yet) upload the data or interact with the spreadsheet, but enough code has been written that a review is a good idea.

nkinkade commented on scraper_test.py:
```diff
@@ -0,0 +1,217 @@
+      files = scraper.download_files('/bin/false', 'localhost/', ['2016/10/26/DNE1', '2016/10/26/DNE2'], '/tmp')
+
+
+  def test_download_files_with_empty_does_nothing(self):
```
There doesn't seem to be any sort of assertion in this test. Is that what you meant?

pboothe replied:
'/bin/false', if run as if it were the binary for rsync, would cause the program to crash and for SystemExit to be called. By not crashing or throwing an exception, this test verifies that empty lists are never passed to rsync.

nkinkade commented:
General style comment. Google's python style guide specifies indention of 4 spaces, but these files appear to only be using a 2 space indention.

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,306 @@
+  """
+  filtered = []
+  for fname in files:
+    if fname.count('/') < 3:
```
Are .gz and .meta the only two kinds of files we upload? I was trying to leave it flexible enough to handle anything, but if it doesn't need to I will happily restrict it to just those two file types.

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,306 @@
+  parser = argparse.ArgumentParser(
+      description='Scrape a single experiment at a site, upload the results '
+                  'if enough time has passed.')
+  parser.add_argument('--rsync_host', metavar='HOST', type=str, nargs=1,
```
Sold! I had no realized that I was confusing things by explicitly specifying the number of arguments. Changed.

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,306 @@
+  try:
+    logging.info("rsync file list discovery from %s to %s", rsync_url,
+                 destination)
+    command = [rsync_binary, rsync_url, destination, '--list-only', '-r']
```
Fixed. This also made me realize that `destination` was not needed in this function at all, and as such it has been removed.

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,306 @@
+    logging.info("rsync file list discovery from %s to %s", rsync_url,
+                 destination)
+    command = [rsync_binary, rsync_url, destination, '--list-only', '-r']
+    logging.info("Listing files on server with the command: %s", command)
```
So I actually struggled with this. It's totally nicer for the reader, but one of the cool things about the `logging` functions is that they don't do any work if the log level isn't high enough (that is: they check the log level and then format the output, instead of the other way around), and performing that `.join` every time would be doing work whether the log level was high enough or not. So I left the output as slightly gross. If it still bugs you after this explanation, let me know and I will change it.

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,306 @@
+      continue
+    year, month, day, _ = fname.split('/', 3)
+    if not (year.isdigit() and month.isdigit() and day.isdigit()):
+      logging.error("Bad filename. Was supposed to be YYYY/MM/DD, but was %s",
```
The linter and formatter smacked my hand about that as well. Fixed.

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,306 @@
+    rsync_binary: The full path to `rsync`
+    rsync_url: The url from which to retrieve the files
+    files: a list of filenames to retrieve
+    distination: the directory on the local host to put the files
```
[:embarrassed face emoji:] 

Fixed.

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,306 @@
+    os.makedirs(destination)
+
+  # Get the file list and then the files from the server.
+  rsync_url = 'rsync://{}:{}/{}/'.format(args.rsync_host[0], args.rsync_port,
```
You are right. Also, I have added rsync_port (and every other arg) to my command-line tests to create a test case which causes this bug to occur. Your comment made me realize I had never tested the parsing and usage of that particular command-line arg.

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,306 @@
+  # Get the file list and then the files from the server.
+  rsync_url = 'rsync://{}:{}/{}/'.format(args.rsync_host[0], args.rsync_port,
+                                         args.rsync_module[0])
+  files = list_rsync_files(args.rsync_binary[0], rsync_url, destination)
```
TIL! I removed the destination, but I did not know that about rsync's default behavior!

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,306 @@
+  for day in find_all_days_to_upload(destination, new_high_water_mark):
+    tgz_filename = create_tarball(destination, day)
+    upload_tarball(tgz_filename)
+    update_high_water_mark(args.spreadsheet[0],
```
Same good catch.

pboothe commented on scraper_test.py:
```diff
@@ -0,0 +1,217 @@
+-rw-r--r--            716 2016/01/06 18:07:37 2016/01/06/20160106T18:07:33.122784000Z_:0.meta
+-rw-r--r--            103 2016/01/06 22:32:01 2016/01/06/20160106T22:31:57.229531000Z_:0.cputime.gz'''
+    patched_subprocess.return_value = serverfiles
+    files = scraper.list_rsync_files('/usr/bin/rsync', 'localhost', '/tmp')
```
`mock` is magic.

What this does is make it so that, no matter what is passed in to `subprocess.check_output()`, the return value will be the specified string. Because '/tmp/ is only used in the args to rsync, and rsync is only called by `subprocess.check_output()`, it is essentially a dummy arg.

pboothe commented on scraper_test.py:
```diff
@@ -0,0 +1,217 @@
+              '2016',
+              '2016/01',
+              '2016/01/06',
+              '2016/01/06/.gz',
```
`remove_older_files` does that. I wanted to keep the `list_rsync_files` simple enough that all it did was list the files that rsync could see.

pboothe commented:
Quote from Google's internal python style guide: Indent your code blocks with 2 spaces.
Quote from Google's external python style guide: Indent your code blocks with 4 spaces.

WHHYYYYYYY??!!?

I'll update everything to 4 spaces, but will also register my grumpiness with the Python style guide maintainers.

nkinkade commented on scraper.py:
```diff
@@ -0,0 +1,348 @@
+
+def main():  # pragma: no cover
+    # TODO(pboothe) end-to-end tests
+    args = parse_cmdline(sys.argv[1:])
```
This may just be a style preference, and if you prefer the current way, then please do ignore this comment! You don't need to pass `sys.argv` to `parse_args`. [From the documentation]:(https://docs.python.org/3/library/argparse.html#parsing-arguments) 

> In a script, parse_args() will typically be called with no arguments, and the ArgumentParser will automatically determine the command-line arguments from sys.argv.

pboothe replied:
Truth. I did it this way so that I could unit-test `parse_cmdline`

anarkiwi replied:
Few comments:

1. It would be good to use a standard library for locking - for example https://pypi.python.org/pypi/fasteners, or LockFile if the python we are going to work with is old.

2. The current internal scraper has a concept of reusing a partial rsync result. It appears here that we start fresh on each rsync poll attempt. I suggest having the temp directory persist until the rsync succeeds. 

3. Right now you don't specify -z for compression, or --bwlimit. I suggest the latter especially to stop a big rsync transfer from potentially stomping on test or other traffic.

pboothe replied:
Changed to use `fasteners`.

The partial rsync data will be saved as long as we don't delete the data between runs.  We only delete data after it has been uploaded (and such deletion is not implemented in this pull request).

Added a global constant RSYNC_ARGS containing `['-4', '-z', '--bwlimit', '10000']`, and now use that constant to help construct the array that specifies the command we execute. Unless you think that the value for `--bwlimit` should be a command-line parameter?

nkinkade commented on scraper.py:
```diff
@@ -0,0 +1,348 @@
+    Args:
+      rsync_binary: the full path location of rsync
+      rsync_url: the rsync:// url to download the list from
+      destination: the directory on the local host to put the files
```
The above "destination" line can be removed now.

pboothe replied:
Done.

nkinkade commented on scraper.py:
```diff
@@ -0,0 +1,348 @@
+        default=1000000000,
+        required=False,
+        help='The maximum number of bytes in an uncompressed tarfile (default '
+        'is 1e9 = 1,000,000,000 = 1 GB)')
```
s/1e9/10e9/

pboothe replied:
1e9 is 1 followed by 9 zeroes. But I'll change the text to eliminate any chance of controversy around this demonstrably confusing way of expressing things.

pboothe commented on scraper.py:
```diff
@@ -0,0 +1,348 @@
+        '--rsync_binary',
+        metavar='RSYNC',
+        type=str,
+        default=['/usr/bin/rsync'],
```
Yes! This is a leftover from the old method. Added a unit test that catches this, and then fixed the problem.

nkinkade commented:
LGTM

pboothe commented:
Changed to use `fasteners`.

The partial rsync data will be saved as long as we don't delete the data between runs.  We only delete data after it has been uploaded (and such deletion is not implemented in this pull request).

Added a global constant RSYNC_ARGS containing `['-4', '-z', '--bwlimit', '10000']`, and now use that constant to help construct the array that specifies the command we execute. Unless you think that the value for `--bwlimit` should be a command-line parameter?

PTAL.
